crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+///
+///   Error:
+///      in responses, an unsigned 16-bit integer containing the extended
+///      RCODE covering TSIG processing.  In requests, this MUST be zero.
+///
+///   Other Len:
+///      an unsigned 16-bit integer specifying the length of the Other Data
+///      field in octets.
+///
+///   Other Data:
+///      additional data relevant to the TSIG record.  In responses, this
+///      will be empty (i.e., Other Len will be zero) unless the content of
+///      the Error field is BADTIME, in which case it will be a 48-bit
+///      unsigned integer containing the server's current time as the
+///      number of seconds since 00:00 on 1970-01-01 UTC, ignoring leap
+///      seconds (see Section 5.2.3).  This document assigns no meaning to
+///      its contents in requests.
+/// ```
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct TSIG {
+    algorithm: Algorithm,
+    time: u64,
+    fudge: u16,
+    mac: Vec<u8>,
+    oid: u16,
+    error: u16,
+    other: Vec<u8>,
+}
+
+/// Algorithm used to authenticate communication
+///
+/// [RFC8945 Secret Key Transaction Authentication for DNS](https://tools.ietf.org/html/rfc8945#section-6)
+/// ```text
+///      +==========================+================+=================+
+///      | Algorithm Name           | Implementation | Use             |
+///      +==========================+================+=================+
+///      | HMAC-MD5.SIG-ALG.REG.INT | MAY            | MUST NOT        |
+///      +--------------------------+----------------+-----------------+
+///      | gss-tsig                 | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha1                | MUST           | NOT RECOMMENDED |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha224              | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha256              | MUST           | RECOMMENDED     |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha256-128          | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha384              | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha384-192          | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha512              | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha512-256          | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+/// ```
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum Algorithm {
```

Comment:
I think we may want to consider naming this `TsigAlgorithm`, as DNSSEC has a different `Algorithm`, and it's possible they may be used in similar contexts.